[FEEDBACK] Picker iOS

sboddeus opened the pull request:
This was tricker than I original anticipated and I have taken a lot of liberties and best guesses with the implementation, so I am definitely looking for some feedback as I am not happy, but I am also a little stuck 🙃

I will add inline comments to this PR to highlight the parts I am unsure about and provide reasons why I made certain decisions that may otherwise appear odd.

Some notes:
- The implementation is inspired by the TableView implementation, even though `UIPickerViewDelegate` and `UIPickerViewDataSource` behave differently to their `UITableView` counterparts.
- The majority of those delegate and data source methods are hidden, unlike `TableView`, because I felt providing the `pickerData` binding reduces the need for them to be exposed. (I did a similar thing for the segment control).
- `UIPickerView` does not allow reloading of individual rows, just individual components or all components. Similarly, row height and column width are specified at the component level. So I created a component descriptor that contains all the row elements as an array. This feels restrictive and I am open to suggestions on how to improve it.

Thanks.

sboddeus committed:
Get picker view working

sboddeus committed:
Finish adding view convertible to picker view

sboddeus committed:
Add more than one column

sboddeus committed:
Add tests

sboddeus committed:
Merge branch 'master' of github.com:mattgallagher/CwlViews into feature/picker-ios

sboddeus commented on Sources/CwlViews/Implementations/iOS/CwlPickerView_iOS.swift:
```diff
@@ -0,0 +1,370 @@
+	static var viewConstructor: PickerViewName<(_ rowSignal: SignalMulti<Binding.ViewDataType>) -> ViewConvertible> { return .name(PickerView.Binding.viewConstructor) }
+
+	// Composite binding names
+	static var rowSelected: PickerViewName<SignalInput<PickerRowAndComponent>> {
```
I am not necessarily sold on only returning the most recently picked element. Which is the default UIPickerView behaviour. It would be nice to return each selected row from the all the components.

sboddeus replied:
I am also only returning the index, vs returning the actual element. Which is the nice behaviour we have on the table view. This means if users of this wrapper want the actual element they have to do that binding on their 'end'.

sboddeus commented:
Another note:
`UIPickerViewDelegate` has three methods for constructing elements, `titleForRow:inComponent`, `attributedTitleForRow:inComponent` and `viewForRow:inComponent`. If you implement `viewForRow...` the others are ignored (because viewForRow has to return a view). Therefore, I implemented only `viewForRow` and if the user of the class specifies a `PickerComponent.ComponentType.text` or `PickerComponent.ComponentType.attributedText` I simply create a view with a label.

sboddeus commented on Sources/CwlViews/Implementations/iOS/CwlPickerView_iOS.swift:
```diff
@@ -0,0 +1,370 @@
+
+		// Delegate
+		open func pickerView(_ pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusing view: UIView?) -> UIView {
+			if let element = components.at(component)?.elements.at(row) {
```
It would be good to somehow introduce view reuse for efficiency. But I got a little stuck with the implementation.

mattgallagher commented:
Yeah, `UIPickerView` is definitely a tricky class to properly adapt to CwlViews since a bit of wrestling is required to apply a data-driven approach onto Cocoa-style delegates.

I think you are using the right approach, though. As you said, `UIPickerView` reloads components but not individual rows, `ArrayMutation` for components and whole arrays for rows seems appropriate. I don't know that you need the `PickerComponentMutation` typealias (just use `ArrayMutation`).

What's the default value for component width and row height? Does iOS choose default values if you return 0?

I don't think I would bother with the `ComponentType` in `PickerComponent`. I would try to implement `titleForRow` and `attributedTitleForRow` (turning `ViewData` into `String?` or `AttributedString?` respectively) as delegate bindings alongside `viewConstructor` so the underlying class's logic is used to determine which implementation is used and how the row is constructed. For this to work though, you won't be able to implement the method on the `Storage` and must use the `Delegate` (this will prevent `UIPickerView` seeing any of the methods, unless they are directly used).

I notice you picked up the `associatedRowInput` from `TableView` (although it's not yet being used). I think this will need to return a tuple of `(SignalInput<ViewData>, String)` (instead of just `SignalInput<ViewData>`) where the `String` is the "identifier" for the view. This should work like the identifier for a table row. To make this work, there will also need to be a `cellIdentifier` delegate binding (used just like `TableView`). From this, you should be able to reuse rows (if the `cellIdentifier` is unchanged) or call the `viewConstructor` if the `cellIdentifier` is changed.

Also like TableView, the `viewConstructor`, `cellIdentifier` and `didSelectRowAndComponent` should all use a `PickerRow<ViewData>` type that is the `ViewData`, the `row` and the `component` (e.g. a replacement for `PickerRowAndComponent` that includes `ViewData`). This will let decisions be made from either the row/component indices or the data.

The `rowSelect` should emit `PickerRow<ViewData>`, addressing one of your concerns about that binding. There could be a separate composite binding that emits an array of `PickerRow<ViewData>` (one for each component).

mattgallagher commented:
I'll be able to look at this on the weekend. I'll see where you're at and take care of any remaining tasks.

sboddeus committed:
Address some PR comments

- Remove commented out code
- Replace RowAndComponent with RowComponentAndData
- Use the underlying class implementaiton to handle `title` vs `attributedTitle` vs `viewForRow`

sboddeus committed:
Update tests

sboddeus commented:
Thank you for the feedback.
I think I was able to make significant improvements, and I definitely have a better picture of the links between the `Storage` and `Delegate` classes of the `BinderDelegateEmbedderConstructor`.

To answer your question, no, it does not appear that UIPickerView will pick reasonable defaults if you return 0 for row height or column width. They will just be set as 0.

Todo:
- I have not yet addressed the view reuse / `identifier` improvements

sboddeus commented on Tests/CwlViewsCatalog_iOS/Sources/PickerView.swift:
```diff
@@ -44,27 +35,23 @@ func pickerView(_ viewState: PickerViewState, _ navigationItem: NavigationItem)
-							Label(.text <-- data)
-						}
+						.title -- { $0.data }
+//						.viewConstructor -- { data in
```
I have tested this and it works, so I left it in as a comment for easy testing purposes so one can see that `.title` also works as expected and when both are set `viewConstructor` is given precedence.

sboddeus committed:
Add composite binding for returning all selected elements

mattgallagher commented:
I've merged the picker view.

I made some significant changes. Views cache correctly. If you provide a `viewIdentifier` then they cache like UITableViewCells with an identifier (otherwise an identifier is automatically provided based on column).

I moved width and rowHeight out of the PickerComponent and into their own delegate bindings. This is because calculating width usually requires access to the UIPickerView to get its width and the only place this is possible is in a delegate binding. With this change, I removed the PickerComponent entirely.

In the catalog, I added a second PickerView so that one can use `viewConstructor` and the other can use `attributedTitle`/`title`.

mattgallagher commented:
I've made these changes fairly hastily. If they don't look correct or you think I've gone in the wrong direction, let me know.

sboddeus commented:
I am happy with the changes you made. They all make sense to me 👍
